lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp

```diff
@@ -399,6 +417,41 @@ void IMDeadCodeElimPass::rewriteModuleSignature(FModuleOp module) {
   numRemovedPorts += deadPortIndexes.size();
 }
 
+void IMDeadCodeElimPass::eraseEmptyModule(FModuleOp module) {
+  // Public modules cannot be erased.
+  if (module.isPublic())
+    return;
+
+  // If the module doesn't have arguments, operations or annotations, we
+  // consider it to be dead.
+  if (liveModules.count(module) || !module.annotations().empty())
+    return;
+
+  // Ok, the module is empty. Delete instances unless they have symbols.
+  LLVM_DEBUG(llvm::dbgs() << "Erase " << module.getName() << "\n");
+
+  InstanceGraphNode *instanceGraphNode =
+      instanceGraph->lookup(module.moduleNameAttr());
+
+  bool existsInstanceWithSymbol = false;
+  for (auto *use : llvm::make_early_inc_range(instanceGraphNode->uses())) {
+    auto instance = cast<InstanceOp>(use->getInstance());
+    if (instance.inner_sym()) {
+      existsInstanceWithSymbol = true;
+      continue;
+    }
+    use->erase();
+    instance.erase();
+  }
+
+  // If there is an instance with a symbol, we don't delete the module itself.
+  if (existsInstanceWithSymbol)
+    return;
```

Comment:
Can we handle this during `markInstance`, and add the target module to the `liveModule` set if the instance has a symbol?